linux-tq: make localversion deterministic

dwagenk opened the pull request:
The localversion string built into the kernel should be deterministic. With OpenEmbedded it is common practice to apply patches to the sources. The commit id of HEAD will differ each time the patching step is rerun. With the previous implementation of the do_set_localversion_tq task this produced non-deterministic commit ids to be used. This is undesireable, because the commit id read from the version string during runtime is not present in the sources. It also caused problems when building out- of-tree kernel-modules if the kernel itself was taken from sstate cache but the kernel-module needed to be rebuild and retriggered the kernels do_patch task (see this issue in meta-freescale, that was due to a similar localversion modification
https://github.com/Freescale/meta-freescale/issues/961).

Rework to produce a deterministic string that will use the commit id given as SRCREV and count the number of patches applied on top.

With e.g. commit id 11aabbcc and 5 patches this will result in a version string
  +g11aabbcc+p5

Related: https://github.com/Freescale/meta-freescale/pull/1694

dwagenk commented:
I'd suggest not merging this for now, there is some discussion in https://github.com/Freescale/meta-freescale/issues/961 about it breaking `AUTOREV`.

tq-steina commented:
Thanks for the Pull Request. But the author is the commit is messed up: `kas <kas@example.com>`
I can fix it myself if you agree.

Now that the meta-freescale Backport has been merged https://github.com/Freescale/meta-freescale/pull/1695. I will pick this one and add a similar change for `sources/meta-tq/meta-tq/recipes-bsp/u-boot/u-boot-tq.inc`

tq-steina commented:
@dwagenk  An additional note: Could you also integrate the change from https://github.com/Freescale/meta-freescale/commit/84c08eb2921c6c35cdfee142356a7ae6f48eb23d on top?

Commit:
linux-tq: fix SRCREV_machine and AUTOREV use cases

Depending on how one uses the Yocto kernel classes the kernels hash
is defined in either "SRCREV_machine" or "SRCREV". If "SRCREV_machine"
is in use, "SRCREV" stays at its bitbake default "INVALID".

If the "SRCREV_machine" or "SRCREV" is set to "AUTOREV" that value is
replaced by "AUTOINC".

If using "SRCREV_machine" and/or "AUTOREV" do_set_localversion_tq fails.

Cope with both use cases.

Adapted from meta-freescale commit 16a356ef0d5d62090e84530e125ea6952994aaeb
by Max Krummenacher <max.krummenacher@toradex.com>

Signed-off-by: Daniel Wagenknecht <dwagenknecht@emlix.com>

Commit:
linux-tq: make localversion deterministic

The localversion string built into the kernel should be deterministic.
With OpenEmbedded it is common practice to apply patches to the sources.
The commit id of HEAD will differ each time the patching step is rerun.
With the previous implementation of the do_set_localversion_tq task this
produced non-deterministic commit ids to be used. This is undesireable,
because the commit id read from the version string during runtime is
not present in the sources. It also caused problems when building out-
of-tree kernel-modules if the kernel itself was taken from sstate cache
but the kernel-module needed to be rebuild and retriggered the kernels
do_patch task (see this issue in meta-freescale, that was due to a
similar localversion modification
https://github.com/Freescale/meta-freescale/issues/961).

Rework to produce a deterministic string that will use the commit id
given as SRCREV and count the number of patches applied on top.

With e.g. commit id 11aabbcc and 5 patches this will result in a version
string
  +g11aabbcc+p5

Related: https://github.com/Freescale/meta-freescale/pull/1694

Signed-off-by: Daniel Wagenknecht <dwagenknecht@emlix.com>

dwagenk commented:
@tq-steina addressed the commit author and imported the additional commit for AUTOINC from meta-freescale. Also fixed the task order, so the localversion task runs after the patch task.

Tested with linux-imx-tq_6.1 and with the same recipe but SRCREV set to `AUTOINC`. Both create the `-tq+gbc136cc9c1a1+p0` localversion string. Adding a patch to SRC_URI increases the patch count as expected.

tq-steina commented:
Thanks a lot for the contribution. Your patches have passed our internal review and integration and are integrated in kirkstone branch. Once they are publishedon github. I'll close this pull request.

Thanks again
